Support Ports on Multisite Domains

ObliviousHarmony opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

This _should_ cover all of the areas that break when using multisite with a port.

Trac ticket: https://core.trac.wordpress.org/ticket/21077

This also addresses https://core.trac.wordpress.org/ticket/52088 since I had to get it working in order to test this PR out locally while developing it.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ObliviousHarmony committed:
Support Port in Site Creation

Rather than sanitizing the domain as if it was a username we should
sanitize it using the character set defined in the `site-new.php` form.

ObliviousHarmony committed:
Prevent Site Info Port Removal

ObliviousHarmony committed:
Allow Ports in Network Setup

ObliviousHarmony committed:
Added Multisite Environment Support

You can now install WordPress as multisite by setting the
`LOCAL_MULTISITE` environment variable to `true`.

ObliviousHarmony committed:
Linting Fixes

ObliviousHarmony committed:
Fixed Multisite Test

sarinajor commented:
<a href='https://capmistoi-6221.weebly.com'>.</a>
<a href='https://capmistoi-6222.weebly.com'>.</a>
<a href='https://capmistoi-6223.weebly.com'>.</a>
<a href='https://capmistoi-6224.weebly.com'>.</a>
<a href='https://capmistoi-6225.weebly.com'>.</a>
<a href='https://capmistoi-6226.weebly.com'>.</a>
<a href='https://capmistoi-6227.weebly.com'>.</a>
<a href='https://capmistoi-6228.weebly.com'>.</a>
<a href='https://capmistoi-6229.weebly.com'>.</a>
<a href='https://capmistoi-6230.weebly.com'>.</a>
<a href='https://capmistoi-6231.weebly.com'>.</a>
<a href='https://capmistoi-6232.weebly.com'>.</a>
<a href='https://capmistoi-6233.weebly.com'>.</a>
<a href='https://capmistoi-6234.weebly.com'>.</a>
<a href='https://capmistoi-6235.weebly.com'>.</a>
<a href='https://capmistoi-6236.weebly.com'>.</a>
<a href='https://capmistoi-6237.weebly.com'>.</a>
<a href='https://capmistoi-6238.weebly.com'>.</a>
<a href='https://capmistoi-6239.weebly.com'>.</a>
<a href='https://capmistoi-6240.weebly.com'>.</a>

ObliviousHarmony commented:
Thanks @JJJ, any suggestions on moving this forward? I didn't get a reply to anyone in #core on Slack and I can't seem to get any replies on the Trac ticket.

swissspidy committed:
Merge branch 'trunk' into fix/multisite-port

desrosj commented:
👋 @ObliviousHarmony thanks for working on this!

Just wanted to acknowledge that I have seen this, but I can't promise that I'll have the capacity to take a look for the next few weeks. I know it's frustrating, but it's not uncommon for it to take months or years for a given change to get the needed feedback and testing. Especially since a few of the contributors previously active on the ticket have no sponsored contribution time at the moment. 

One thing I recommend for clarity's sake. If you could go through the existing concerns and just briefly document your thinking around how you solved them, that would help someone dive in looking at the ticket history.

desrosj commented on .env:
```diff
@@ -53,6 +53,9 @@ LOCAL_DB_TYPE=mysql
 LOCAL_DB_VERSION=5.7
 
+# Whether or not to enable multisite.
+LOCAL_MULTISITE=false
```
With this now available, would it make sense to turn this to `true` for the multisite test jobs for PHPUnit?

If not, then we should at least have some type of test that ensures the environment starts and works when multisite is enabled. That could potentially be a separate workflow that tests all of the related local environment commands, and perhaps better off as a separate ticket.

ObliviousHarmony replied:
It's worth keeping in mind that PHPUnit tests create their own tables and `wp-config.php` file. The environment variable just switches `install` to `multisite-install` and it wouldn't have any effect on the PHPUnit tests.

It's _probably_ not worth creating a test for installing since all it does is run `multisite-install` instead of `install`. The rewrite rule change is covered by the existing test workflows and the parsing is covered by PHPUnit tests.

ObliviousHarmony committed:
Removed Unnecessary Subdomain Port Removal

It's been months and I'm not totally sure why I left this here in the first place. Subdomains can have ports too, obviously!

ObliviousHarmony committed:
Merge branch 'trunk' into fix/multisite-port

ObliviousHarmony committed:
Replaced Hard-Coded Test Domains

There were a number of tests that failed when `WP_TESTS_DOMAIN`
has a port. This commit fixes all of those tests so that they will pass
with both bare hostnames and those with ports.

To discover these problems I added a port to my local
`wp-tests-config.php` and ran the test suite.

ObliviousHarmony committed:
Added Support For `.env` `WP_TESTS_DOMAIN`

In order to make it easily possible to run the test suite with
a different domain this new environment variable
changes what's set in the tests config. This is
good for things like testing it with ports.

ObliviousHarmony committed:
Added Missing Multisite oEmbed Port

Previously, the port was not being used when requesting
the site for multisite installations. This breaks oEmbed
when a port is used.

ObliviousHarmony committed:
Added Missing RSS Link Port

The self link should include the port instead of just
using the host. This way, it will link to the correct
site.

ObliviousHarmony committed:
Check Port When Securing Calculated `srcset`

We only set the protocol to `https://` when the url is
for the current request. Since `HTTP_HOST`
will include a port we need to consider that too.

ObliviousHarmony committed:
Added `WP_TESTS_DOMAIN` To GitHub Workflows

Not only does this let us test this PR in CI but it also ensures
that future PRs will still work with ports in the
test domain.

ObliviousHarmony reviewed (commented):
Thanks for taking a peek @desrosj,

Since it's been a few months I decided to take another pass and make _absolutely sure_ this change works as expected. A big part of that was buttoning up the use of `WP_TESTS_DOMAIN` in the PHPUnit suite so that it can be ran with a port in `WP_HOME`. In the process, I even found a few really small cases where WordPress breaks when on a host using a non-standard port :smile:

> One thing I recommend for clarity's sake. If you could go through the existing concerns and just briefly document your thinking around how you solved them, that would help someone dive in looking at the ticket history.

Really, the only concern in the ticket from folks is uncertainty around what aspects of multisite might break with a port present in the domain. There are, however, also comments making similarly minimalistic changes as I have and reporting that everything works correctly.

@spacedmonkey noted two years ago that "adding a port to the domain breaks things all over WordPress" but that hasn't been my experience. Even WordPress' own development environment uses a non-standard HTTP port and that use-case seems completely supported.

The only potential issue I see with storing the port in the domain for the site is that `WP_Site_Query` now requires the port if one is set. This wouldn't impact any existing multisite installations and I _seriously_ doubt many users are creating production sites with a port in the first place.

In my case, I'd like to support ports so that we can shard E2E tests using a single environment. This is great for running tests more quickly without having to spin up a bunch of environments. The same command can be used both in CI and locally with much better performance. It's also super useful for isolating groups of tests due to them affecting some global state.

johnbillion committed:
Merge branch 'trunk' into fix/multisite-port

johnbillion committed:
Let's see if we can get this workflow running.

johnbillion committed:
Add the test domain to the job name when its value is not the default.

johnbillion commented:
@ObliviousHarmony I pushed a couple of changes to get the PHPUnit tests working on this PR. There's [one failing job](https://github.com/WordPress/wordpress-develop/actions/runs/8558389043/job/23455056761) which is the newly introduced job that tests a single site installation with a domain name that includes a port number. The failure is only caused by the `wp-api-generated.js` fixture getting updated. Not sure on the best fix for that.

ObliviousHarmony commented:
Thanks @johnbillion! Yeah, that's definitely tricky. I wouldn't want to do anything as broad as removing the port before writing the file but I don't see any real alternative. Modifying the test fixture this generates feels like a bad idea anyway. How would you feel about expanding the action's step to allow for certain differences in certain cases? The matrix is pretty extensive and I don't know that an exception in this one case would be that bad.

I added this test because (as this PR shows) there are a _ton_ of tests that break with a port that needed updating. It seems wise to have a test that uses a port to make sure everything works correctly.  In the interest of moving this forward, however, it might make sense to pull that test out. The tests pass and so I think that's a good indicator for this PR at least. The problem

johnbillion committed:
Merge branch 'trunk' into fix/multisite-port

johnbillion committed:
Merge branch 'trunk' into fix/multisite-port

ObliviousHarmony commented on .github/workflows/phpunit-tests.yml:
```diff
@@ -36,7 +36,7 @@ jobs:
   test-with-mysql:
     name: PHP ${{ matrix.php }}
-    uses: WordPress/wordpress-develop/.github/workflows/phpunit-tests-run.yml@trunk
+    uses: ObliviousHarmony/wordpress-develop/.github/workflows/phpunit-tests-run.yml@fix/multisite-port
```
Just noting @johnbillion that this needs to be changed back before merging so that it doesn't break the workflow. We don't need to run actions using a random contributor's branch :smile:

johnbillion replied:
Thanks :D

johnbillion committed:
Skip running the REST API client fixture tests when using a non-default test domain name.

johnbillion committed:
Account for port when changing a site path.

johnbillion reviewed (approved):
I pushed two commits:

* 0665ca4 which skips the REST API client fixture test when the tests run with a port in the host name, I couldn't think of any other way around this.
* 49365cb which ensures the `home` and `siteurl` options are updated with the new value when the path is updated from the site editing screen.

I've reviewed everything else in this PR and it looks good.

ObliviousHarmony commented:
Thanks for the review and helping get this across the finish line @johnbillion, I've been pretty swamped this week :smile:

- https://github.com/WordPress/wordpress-develop/commit/0665ca4a7b66536c2d60c61bd43ac53b5d5076ca: Checking for `example.org` like this seems like a hack at first but given the purpose of this test's output I don't think there's a problem.
- https://github.com/WordPress/wordpress-develop/commit/49365cb8356e067a9807fd45187e1225982bed77: Nice catch!

It'll be nice to get this across the finish line. I'm planning to use this to run E2E test shards in WooCommerce without needing to spin up a bunch of identical environments. Faster tests both locally and in CI :smile:

johnbillion commented:
https://core.trac.wordpress.org/changeset/58097
